### PR TITLE
[SPARK-34521][PYTHON][SQL] Fix spark.createDataFrame when using pandas with StringDtype

### PR DESCRIPTION
**Pull request by nicolasazrak**
### What changes were proposed in this pull request?

This change fixes `SPARK-34521`. It allows creating a spark DataFrame from a pandas DataFrame that is using a `StringDtype` column and arrow pyspark enabled.

### Why are the changes needed?

Pandas stores string columns in two different ways: using a numpy `ndarray` or using a custom `StringArray`. The `StringArray` version is used when specifing the `dtype=string`. When that happens, spark cannot serialize the column to arrow. Converting the `Series` before fixes this problem. 

However, due to the different ways to handle string columns, doing `spark.createDataFrame(pandas_dataframe).toPandas()` might not equal to `pandas_dataframe`. The column dtype could be different.

More info: https://pandas.pydata.org/docs/user_guide/text.html

### Does this PR introduce _any_ user-facing change?

Trying to create a spark `DataFrame` from a pandas `DataFrame` using a string dtype and `"spark.sql.execution.arrow.pyspark.enabled"` now doesn't throw an exception and returns the expected dataframe.

Before:

`spark.createDataFrame(pd.DataFrame({"A": ['a', 'b', 'c']}, dtype="string"))`
```
Error
Traceback (most recent call last):
  File "/home/nico/projects/playground/spark/python/pyspark/sql/tests/test_arrow.py", line 415, in test_createDataFrame_with_string_dtype
    print(self.spark.createDataFrame(pd.DataFrame({"A": ['a', 'b', 'c']}, dtype="string")))
  File "/home/nico/projects/playground/spark/python/pyspark/sql/session.py", line 823, in createDataFrame
    return super(SparkSession, self).createDataFrame(  # type: ignore[call-overload]
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/conversion.py", line 358, in createDataFrame
    return self._create_from_pandas_with_arrow(data, schema, timezone)
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/conversion.py", line 550, in _create_from_pandas_with_arrow
    self._sc  # type: ignore[attr-defined]
  File "/home/nico/projects/playground/spark/python/pyspark/context.py", line 611, in _serialize_to_jvm
    serializer.dump_stream(data, tempFile)
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/serializers.py", line 221, in dump_stream
    super(ArrowStreamPandasSerializer, self).dump_stream(batches, stream)
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/serializers.py", line 81, in dump_stream
    for batch in iterator:
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/serializers.py", line 220, in <genexpr>
    batches = (self._create_batch(series) for series in iterator)
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/serializers.py", line 211, in _create_batch
    arrs.append(create_array(s, t))
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/serializers.py", line 185, in create_array
    raise e
  File "/home/nico/projects/playground/spark/python/pyspark/sql/pandas/serializers.py", line 175, in create_array
    array = pa.Array.from_pandas(s, mask=mask, type=t, safe=self._safecheck)
  File "pyarrow/array.pxi", line 904, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 252, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 107, in pyarrow.lib._handle_arrow_array_protocol
ValueError: Cannot specify a mask or a size when passing an object that is converted with the __arrow_array__ protocol.

```

After:

`spark.createDataFrame(pd.DataFrame({"A": ['a', 'b', 'c']}, dtype="string"))`
> `DataFrame[A: string]`

### How was this patch tested?

Using the `test_createDataFrame_with_string_dtype` test.

### CODE REVIEW

**Review by itholic — APPROVED** [review]
LGTM

**Comment by itholic** [comment]
Not a big deal, but could you add a simple example to PR description what issue is resolved with before/after ??

e.g.

Before:
```python
>>> spark.createDataFrame(pd.DataFrame({"A": ['a', 'b', 'c']}, dtype="string"))
.../spark/python/pyspark/sql/pandas/conversion.py:402: UserWarning: createDataFrame attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Cannot specify a mask or a size when passing an object that is converted with the __arrow_array__ protocol.
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.
  warnings.warn(msg)
DataFrame[A: string]
```


After:
```python
>>> spark.createDataFrame(pd.DataFrame({"A": ['a', 'b', 'c']}, dtype="string"))
DataFrame[A: string]
```

**Comment by nicolasazrak** [comment]
@itholic Thanks. I've updated the description to reflect the change.

**Comment by nicolasazrak** [comment]
Hey @itholic can we merge this?

**Comment by itholic** [comment]
It's LGTM but let's wait until other members to verify this change since I don't have permission to merge 😅

cc @HyukjinKwon @ueshin Could you take a look at this one when you find some time ?

File: python/pyspark/sql/pandas/serializers.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/serializers.py:173]: so are you saying that the type of strings from pandas can produce different type in arrow? cc @BryanCutler

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/serializers.py:173]: while I understand that we can work around,

> Pandas stores string columns in two different ways: using a numpy `ndarray` or using a custom `StringArray`. The `StringArray` version is used when specifing the `dtype=string`. When that happens, spark cannot serialize the column to arrow.

This sounds like an issue in Arrow side.

Comment (nicolasazrak) [reply python/pyspark/sql/pandas/serializers.py:173]: Now that I think it again, adding support for the `StringArray` in arrow is the real solution, this change is just a workaround. I don't know much about the arrow internals or if they can add some metadata to support two different string types. If you feel this is better done in arrow upstream I can close the PR and investigate it from the arrow side. Otherwise, we can leave this patch adding a comment and keep investigating in arrow.

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/serializers.py:173]: I'll leave it to @BryanCutler (whos a maintainer for both PySpark and PyArrow).

Comment (ueshin) [reply python/pyspark/sql/pandas/serializers.py:173]: I'm afraid I don't think this is a right fix.

Actually this happens with `Int64`, `Int32`, or other extension dtypes.

```py
>>> spark.createDataFrame(pd.DataFrame({"A": [1, 2, 3]}, dtype="Int64"))
.../python/pyspark/sql/pandas/conversion.py:422: UserWarning: createDataFrame attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Cannot specify a mask or a size when passing an object that is converted with the __arrow_array__ protocol.
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.
  warnings.warn(msg)
DataFrame[A: bigint]
```

The error message says:

```
Cannot specify a mask or a size when passing an object that is converted with the __arrow_array__ protocol.
```

, so we should follow the message.

How about at line 163,

```py
if hasattr(s.values, '__arrow_array__'):
    mask = None
else:
    mask = s.isnull()
```

Comment (BryanCutler) [reply python/pyspark/sql/pandas/serializers.py:173]: I agree with @ueshin , the correct thing to do would be to continue to use pyarrow to convert the column in this case with `mask = None`, that way the implementation is sure to do the conversion most efficiently. I think for this case it should then produce a standard arrow UTF-8 array that PySpark could handle.

Comment (nicolasazrak) [reply python/pyspark/sql/pandas/serializers.py:173]: @ueshin Makes a lot of sense. I wasn't aware this was a problem for other dtypes. Removing the mask works, but writing a quick test:

```python
with self.sql_conf({"spark.sql.execution.arrow.pyspark.enabled": True}):
    pandas_df = pd.DataFrame({"col": [1, 2, 3, None]}, dtype="Int64")
    df = self.spark.createDataFrame(pandas_df)
    assert_frame_equal(pandas_df, df.toPandas())
```
fails with:
```
Attribute "dtype" are different
[left]:  Int64
[right]: float64
```

I'm going to take a look where the conversion is failing.

Comment (ueshin) [reply python/pyspark/sql/pandas/serializers.py:173]: I think it's fine if the `df`'s data type is `LongType` for `dtype="Int64"` because `df.toPandas()` won't keep the extension dtypes.
So the dtype of `df.toPandas()` will be `int64`, or `float64` if containing `None`.

**Comment by ueshin** [comment]
ok to test

**Review by ueshin — COMMENTED** [review]
LGTM, except for the lint-python error.
@nicolasazrak Could you run `./dev/reformat-python` to fix the linter error?
@BryanCutler Could you double check the fix is right? I'd leave it to you.

File: python/pyspark/sql/pandas/serializers.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/serializers.py:218]: this actually reminds me of https://github.com/apache/spark/pull/28743 how is it related to each other?

Comment (nicolasazrak) [reply python/pyspark/sql/pandas/serializers.py:218]: Looks really similar. I've added my tests to that branch and they passed. (However, lot of other tests failed but probably for other reasons).

**Comment by BryanCutler** [comment]
The difference with #28743 is that was trying to deal with pyarrow extension types. For a pandas extension type the `__arrow_array__` interface will return an arrow array which could be a standard arrow type or an extension type. For this PR, we are talking about a standard string array, which pyspark can work with. Otherwise, if it's a pyarrow extension type, the storage type would need to be checked, which would be a standard arrow type. From that, pyspark could work with the storage type but that might not be very useful because all of the extension would be stripped out.

This PR is a step in the right direction, so I think it's ok to merge. This will add support for any pandas extension types that are backed by a standard arrow array, although I don't think it will be able to convert it back to pandas as the original extension type. To fully support pandas/pyarrow extension types we would need to propagate the extension type info through spark so that when it is worked on again in python, the extension part can be loaded back up. I'm not exactly sure how difficult that might be to do.

File: python/pyspark/sql/pandas/serializers.py
Comment (BryanCutler) [reply python/pyspark/sql/pandas/serializers.py:218]: I think it's preferred to use `s.array` instead of `s.values`, would you mind giving that a try?

**Review by BryanCutler — COMMENTED** [review]
Looks fine to me, I just had a minor request to use `s.array` instead.

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Thanks @BryanCutler !

**Comment by BryanCutler** [comment]
merged to master, thanks @nicolasazrak . Do you have an apache.org account so I could assign the issue to you?

**Comment by nicolasazrak** [comment]
@BryanCutler Yes, my username is `nicolasazrak`

**Pull request closed without merging**
